aten/src/ATen/core/thread_pool.cpp

```diff
@@ -71,12 +71,25 @@ void ThreadPool::workOnTasksUntilCompleted(
     return;
   }
   std::condition_variable finished;
-  future->addCallback([&] { finished.notify_all(); });
+  bool fired = false;
+  future->addCallback([&] {
+    std::unique_lock<std::mutex> future_lock(future->get_mutex());
```

Comment:
I think we need some better documentation on ivalue, documenting the relationship between the locks and when callbacks are called. (This appears to be correct but I had to check the implementation of `addCallback` to see if callbacks are run with or without the lock.)